src/rust/src/wallet.rs

```diff
@@ -135,33 +139,80 @@ impl Wallet {
         Wallet {
             key_store: KeyStore::empty(),
             wallet_tx_notes: HashMap::new(),
+            nullifiers: BTreeMap::new(),
             witness_tree: BridgeTree::new(MAX_CHECKPOINTS),
             last_observed: None,
             locked_notes: HashSet::new(),
             spent_notes: HashSet::new(),
         }
     }
 
-    pub fn checkpoint_witness_tree(&mut self) {
-        self.witness_tree.checkpoint();
+    pub fn checkpoint(&mut self, block_height: BlockHeight) -> bool {
+        if let Some(last_observed) = &self.last_observed {
+            if last_observed.height == block_height {
+                self.witness_tree.checkpoint();
+                true
+            } else {
+                false
+            }
+        } else {
+            // it's fine to checkpoint the empty tree.
+            self.witness_tree.checkpoint();
+            true
+        }
     }
 
-    pub fn rewind_witness_tree(&mut self) -> bool {
-        self.witness_tree.rewind()
+    pub fn rewind(&mut self) -> bool {
+        if let Some(last_observed) = self.last_observed.as_mut() {
+            // when we rewind, we need to remove the spentness marks
+            // from the nullifiers we're tracking
+            if self.witness_tree.rewind() {
+                last_observed.height = last_observed.height - 1;
+                last_observed.block_tx_idx = 0;
```

Comment:
Hmm, this isn't correct unless the previous block only had a coinbase transaction. We don't remember how many transactions were appended in prior blocks, and thus cannot set this correctly.

While reviewing this PR, I also noticed that we don't even update `self.last_observed` in `Wallet::append_bundle_commitments`; we just check its (immediately outdated) value.

I think that `last_observed.block_tx_idx` needs to be `Option<usize>`, and then we build simple state transitions from the data we _do_ see:
- Appending:
  - If `last_observed.height == block_height && last_observed.block_tx_idx== Some(block_tx_idx - 1)`:
    - Append bundle.
    - Increment `last_observed.block_tx_idx`.
  - Else if `block_height == last_observed.height + 1 && block_tx_idx == 0`:
    - Append the bundle.
    - Increment `last_observed.height`.
    - Set `last_observed.block_tx_idx` to `Some(0)`.
  - Else append is invalid, return error.
- Rewinding:
  - Decrement `last_observed.height`
  - Set `last_observed.expected_next_tx_idx` to `None`.

After a rewind, the next appended tx will be in one of several possible states:
-  It is the first transaction in the next block, in which case we don't directly check `last_observed.block_tx_idx` but _do_ directly constrain its logical next value `0`.
- It is a transaction attempting to extend the previous block, in which case the check fails (because `None != Some(block_tx_idx - 1)`), which is what we want because after a rewind we should have a complete block (i.e. the node should never start a new block before finishing the previous block).
- It is some other transaction, in which case we error like usual.

Thus setting `last_observed.expected_next_tx_idx` to `None` ensures that the only possible valid subsequent transitions are either another full-block rewind, or an append of the first transaction in the next block.

Reply:
It occurs to me that there might be another problem even with block index equality, which is that if a block contains no transactions with Orchard bundles that we might never even be called, and so the block height check needs to be `block_height > last_observed.height` - essentially, we just need to ensure that these values are always monotonically increasing, but we can't do anything finer-grained than that.